multicluster/service-mirror/cluster_watcher.go

```diff
@@ -1235,13 +1247,13 @@ func (rcsw *RemoteClusterServiceWatcher) createEndpointMirrorService(ctx context
 		return nil, err
 	}
 
-	endpointMirrorAnnotations := map[string]string{
-		// needed to detect real changes
-		consts.RemoteResourceVersionAnnotation: resourceVersion,
-		consts.RemoteServiceFqName:             fmt.Sprintf("%s.%s.%s.svc.%s", endpointHostname, exportedService.Name, exportedService.Namespace, rcsw.link.TargetClusterDomain),
-	}
+	endpointMirrorAnnotations := rcsw.getMirroredServiceAnnotations(exportedService)
+
+	// needed to detect real changes
+	endpointMirrorAnnotations[consts.RemoteResourceVersionAnnotation] = resourceVersion
+	endpointMirrorAnnotations[consts.RemoteServiceFqName] = fmt.Sprintf("%s.%s.%s.svc.%s", endpointHostname, exportedService.Name, exportedService.Namespace, rcsw.link.TargetClusterDomain)
 
-	endpointMirrorLabels := rcsw.getMirroredServiceLabels()
+	endpointMirrorLabels := rcsw.getMirroredServiceLabels(exportedService)
```

Comment:
I think it would be better here if we don't copy over the service annotations and labels. I'd leave the annotation as-is and then use `nil` for `rcsw.getMirroredServiceLabels()`.

This logic comes from the new StatefulSet feature. Here, we are essentially creating a "synthetic" clusterIP service for each pod with a hostname that sits behind a headless service. So, for example, if we have:
```
statefulset={pod-0}, headless-svc={clusterIP: None}, headless-ep={(pod-0, pod-0-ip)}
=> mirrors
headless-svc-mirror={clusterIP: None}, headless-svc-pod-0={clusterIP}, headless-svc-mirror-ep={(pod-0, headless-svc-pod-0-IP)}
```

Essentially, those `endpointMirrors` are created from an endpoint, a pod. I don't think we should copy over the annotations and labels from the "root" service as that would just clutter the service for this instance with metadata we might not need, or that might not apply. These labels and annotations are present on the root service anyway so they can be easily accessed through that.

We should instead opt to keep these as lightweight and "decision-free" as possible, they're only there to route to a specific instance between clusters. Think the default labels and annotations we used to put on a service would suffice here 🤔  @importhuman lmk if this makes sense.

Reply:
I didn't know about headless services and endpoints, so I tried to do some reading. As far as I've understood, headless services are those that don't have a clusterIP (clusterIP: none), and endpoints are the objects through which a service connects to pods. 
What I had tried was to mirror annotations and labels in the cases where a mirror service was being created/updated (since the original issue wanted to copy all metadata from original service), but not for other types such as endpoints. Here, I saw a *service mirror* for endpoints, so I copied the annotations & labels. 
However, if I've understood things correctly, endpoints (and endpoint service) wouldn't be used directly, so what you're suggesting should be alright.

Please let me know if I've interpreted things correctly, and if I should or shouldn't copy over labels/annotations.

Reply:
Yup, the way you understand it is correct.

> What I had tried was to mirror annotations and labels in the cases where a mirror service was being created/updated (since the original issue wanted to copy all metadata from original service), but not for other types such as endpoints. Here, I saw a service mirror for endpoints, so I copied the annotations & labels.

Yeah, so you're right here too. We want to copy over the labels and annotations from the original service to the mirror service. The endpoint mirrors are synthetic though -- they are services we create so that clients can connect directly to a pod belonging to an original service. They _would_ be used directly by a client, but these synthetic services should be thought of more as a pod replacement and not as a service in and of itself.

I think it's easier if we exemplify this a bit. You'll find the manifests for a statefulset and a headless service in the spoiler tag below.

<details> <summary> statefulset manifest </summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-svc
  labels:
    mirror.linkerd.io/exported: "true"
spec:
  selector:
    app: nginx-ss
  clusterIP: None
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: nginx-ss
spec:
  serviceName: nginx-svc
  selector:
    matchLabels:
      app: nginx-ss
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx-ss
    spec:
      containers:
      - name: nginx
        image: nginx
```

</details>

If we apply this in a cluster called `west`, we will have a statefulset with 3 instances (nginx-0, nginx-1, nginx-2) and a headless service. Through the headless service, clients can connect directly to any of the instances by using the pod DNS entry the service creates, i.e `nginx-0.nginx-svc.default.svc.cluster.local`. If we mirror this headless service in a cluster called `east`, we will want to preserve this functionality. Let's say our mirror for the headless service is called `nginx-svc-west`. We would expect to be able to connect to it through `nginx-0.nginx-svc-west.default.svc.cluster.local` -- however, to configure DNS we need an IP address to be associated with this entry. What we do is create a clusterIP service for each instance, these are the `endpointMirrors`. In `west`, our endpoints object has a hostname and an IP for each pod in this statefulset, in `east` we don't have pods so we create a clusterIP service instead to have a synthetic IP. Each synthetic service will also point to the gateway.


Because of this, I don't think we should copy over the annotations. The annotations and labels will exist on the root headless service, if there is a need for it, a client can get it off the service. I think we shouldn't convolute those synthetic services (i.e the `endpointMirrors`), they should only serve to route the traffic to a pod instance. Since the root service is so closely coupled with all services, there would still be access to the labels and annotations, these should imo only be on the `headlessServiceMirror` though.

Hope this makes sense!

Reply:
This was extremely helpful, thank you!